src/haproxy/conn.go

```diff
@@ -0,0 +1,126 @@
+package haproxy
+
+import (
+	"bufio"
+	"net"
+	"time"
+
+	proxyproto "github.com/pires/go-proxyproto"
+)
+
+// Conn is a `net.Conn` compatible struct that handles PROXY header checking.
+type Conn struct {
+	rd  *bufio.Reader
+	c   net.Conn
+	r   net.Addr
+	l   net.Addr
+	hdr *proxyproto.Header
+}
+
+// NewConn wraps a connection with a `bufio.Reader` for checking if the
+// PROXY protocol headers are present
+func NewConn(nc net.Conn) (net.Conn, error) {
+	c := &Conn{
+		c:  nc,
+		rd: bufio.NewReader(nc),
+	}
+	if err := c.ProxyInit(); err != nil {
+		return nil, err
+	}
+	return c, nil
+}
+
+// ProxyInit checks for the PROXY protocol headers, and if not returns
+// the connection unmolested.
+func (c *Conn) ProxyInit() error {
+	pc, err := proxyproto.Read(c.rd)
+	if err != nil &&
+		err != proxyproto.ErrNoProxyProtocol &&
+		err != proxyproto.ErrInvalidLength {
+		return err
+	} else if err == proxyproto.ErrInvalidLength {
+		// Failed to parse header, the connection is probably about to drop anyway,
+		// it's not a PROXY header at least.
+		return nil
+	}
```

Comment:
This error handling code is a little difficult to follow. It could be improved by having a branch fo each condition, or perhaps even using a switch.  It wasn't immediately obvious to me that `nil` was returned when `proxyproto.Read()` returns `proxyproto.ErrNoProxyProtocol`.

Reply:
replace with a switch and better comments.